### PR TITLE
Progression History

### PR DESCRIPTION
**Pull request by MattEqualsCoder**
Pretty rough draft of changes, but wanted to get this out there somewhat for initial general implementation. Need to do some documenting and clean up.

### CODE REVIEW

**Comment by MattEqualsCoder** [comment]
Did some testing, and it seems to work as expected. Here's an example log that was created:

```
SMZ3 Cas’ run progression log
=============================
Generated on Friday, July 8, 2022 12:13:56 AM
Seed: 1988979802
Settings String: 6wYAAB+LCAAAAAAAAAp1lVnTqjgQhv8Lt35VbC5o1VwAgoqgKLhOzUWEIAgmEBK3U+e/T/DDObPUcEOnu3m7SeiHH8IEXKGHYyiMpC/hqLr4nEVvO/B+2TMKry6OAM0wqoXRD8En+ExgXWc3GNwxid9ZHiZl+rYMfD3VrYXptxWUIIIOu5Z89ZMvU5BTjItG2uf3d44PzztMivgfzgkG38YcPoOUJUnx3eya6wmjBBQ1/BLGWQ1OBQxKnBWQ8M7/HQkJiHJI2oT6f8LTDNFfsQlAGM3QLUNRxnOEkfwlWA8KUQxjr2YBK0tMqDCihPHstrcxQ2eIkcfqdvOmEBBq4gITYaR+CS6+c09BUwPCMighjN/utg8etRAk52cT/asRDyLWpvIWTFAzUASeD2iUwvpTfwL5g4DCgOctUfH862k3Q/mCH7MwEg64TjOB9wqurG59a1CnjStD5wK+t/+j6LGCZq2n1eJnUAOU0V/qTTmuIg81bTgYapLSaEFKuVwdUMJvPNpZebpudFxT/1ylvNEoU176SuSvmkSXi2QF6nLWjal8IFd9Z1KWprO+lN8v671zjB+XYCyq1dZz9srZecrqeuBdXiw9kp3pO+EOZ2R57KCqh6PB5tBXXaqSBVsMV+Pda2ZsN1qllWHvZVlaN+yv3O5xE+x7eN0hSyglq35/FvY2EHsnNQrFWX+zW/rweXiapyoou5eydtn59KDiNWOmNbCN16vy+Wb7BxSbpVbtiG4clTLVLw8g694xk/sb8aX2BvMpxFFH1xzxcdtPtUXBjNkgVkLttF1fzGrpdErW63iknlzH9nR6CMZ+ddUTLia7wCW5fT+uBn0N3kOfKqFTdZxneTlK+WxQJLjb7euTxySmx1R63aY+Ydc5uK3yUvI2C/ueKZZpg729H8eH3kUhsgEN60CZJS6nhZioaaZ2nmelkpwhjvWTsy3TqeSzUzCVqt02OqoTTcTXx8U5LgbXFdMCU/fE1wTRg/MUkSzHe5l1GKm7NyuNHqb9Wg8guJjrHbUXd3Waur1HmDhMz9VTifVCk/2ZY67V7lJ6oMFwGD4Mp2cqPRkFyF7OwiUF8kPzFvUrw9VunyobJZqcJhtkH6WV0zGrTWLwj1shniUX0KIiTidepUzLab7c5vP9am3E6jZZhY4hX/ZLSwzXSXAduq+Vq51ytqFLQHYHumIB9XaS6Ly6i/kN13nWd6ykHOvzxeRAjJNcdvahCE7iXEQ9VxQn8Vjx+bnPVXOYphIlKZuqyVibbYfV0JvIBFzsDXJpFya5bUix1PHzMf/Af+NzYOLy2cyHjuKGVp+x+MzXB6kN7hqscixagBTPdv37H00KZ7CJUZKdv7kLbxDRICLwrlPKmRXghHKd/KPZZticzD6+Q9KgOHiPaFsTJjQAKPYzuoYVyzjFg7KZUgMUHLNtEmAoSn0Qf1JmETQguH7iNvetcTwGJF9j3PTawmDGG+V0gE3Vb9a3AZ+/F+b05Z3wX0JNIYfhB268++/iH6I0f5SGYP/Vd7NzSt9ECjCjqQ1yaBdZWf6d6Duu1dQeZ0mSRRxhnFXKz59/AqIJlzPrBgAA

[00:00:00] Entered Central Crateria
[00:00:28] Entered Light World South
[00:00:49] Entered Light World North West
[00:03:09] Picked up a Gloves from Light World North West - Blind's Hideout - Top
[00:04:07] Picked up a Gloves from Light World North West - Kakariko Well - Middle
[00:04:08] Picked up the Moon Pearl from Light World North West - Kakariko Well - Right
[00:04:32] Defeated Ridley
[00:04:36] Defeated Armos Knights

Final time: 00:04:42
```

**Comment by Vivelin** [comment]
You defeated Ridley and Armos Knights within 4 seconds of each other? That's some impressive tech

File: src/Randomizer.App/Windows/RomListWindow.xaml.cs [hunk 1]
```diff
@@ -594,5 +605,39 @@ private void CopyTextToClipboard(string text)
...
+            if (rom.TrackerState == null)
+            {
+                MessageBox.Show(this, "There is no history for the selected rom.",
+                        "SMZ3 Cas’ Randomizer", MessageBoxButton.OK, MessageBoxImage.Warning);
+                return;
+            }
+
+            _dbContext.Entry(rom.TrackerState).Collection(x => x.History).Load();
+
+            if (rom.TrackerState.History == null || rom.TrackerState.History.Count == 0)
```
Comment (Vivelin) [inline src/Randomizer.App/Windows/RomListWindow.xaml.cs:626]: Would it be possible to disable the option if you right-click on a rom without history?

File: src/Randomizer.SMZ3.Tracking/HistoryService.cs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+namespace Randomizer.SMZ3.Tracking.VoiceCommands
+{
+    /// <summary>
+    /// Service for managing the history of events through a playthrough
+    /// </summary>
+    public class HistoryService : IHistoryService
+    {
+        ILogger<HistoryService> _logger;
+        Tracker? _tracker;
+        List<TrackerHistoryEvent>? _events;
```
Comment (Vivelin) [inline src/Randomizer.SMZ3.Tracking/HistoryService.cs:23]: These are missing access modifiers

File: src/Randomizer.SMZ3.Tracking/Tracker.cs [hunk 3]
```diff
@@ -825,29 +844,65 @@ public void ConnectToChat(string? userName, string? oauthToken, string? channel,
...
         /// </summary>
-        public virtual void StartTimer()
+        public virtual void StartTimer(bool isInitial = false)
         {
+            _undoStartTime = _startTime;
             _startTime = DateTime.Now;
+
+            if (!isInitial)
+            {
+                Say("Timer resumed");
```
Comment (Vivelin) [inline src/Randomizer.SMZ3.Tracking/Tracker.cs:854]: Should make this response configurable.

File: src/Randomizer.SMZ3.Tracking/Tracker.cs [hunk 4]
```diff
@@ -825,29 +844,65 @@ public void ConnectToChat(string? userName, string? oauthToken, string? channel,
...
+            _undoSavedTime = SavedElapsedTime;
+            _undoStartTime = _startTime;
+
             SavedElapsedTime = TimeSpan.Zero;
-            StartTimer();
+            _startTime = DateTime.Now;
+
+            if (!isInitial)
+            {
+                Say("Timer reset");
```
Comment (Vivelin) [inline src/Randomizer.SMZ3.Tracking/Tracker.cs:872]: Should make this response configurable.

File: src/Randomizer.SMZ3.Tracking/Tracker.cs [hunk 5]
```diff
@@ -825,29 +844,65 @@ public void ConnectToChat(string? userName, string? oauthToken, string? channel,
...
         /// </summary>
         public virtual void PauseTimer()
         {
+            _undoSavedTime = SavedElapsedTime;
+            _undoStartTime = _startTime;
+
             SavedElapsedTime = TotalElapsedTime;
             _startTime = DateTime.MinValue;
+
+            Say("Timer paused");
```
Comment (Vivelin) [inline src/Randomizer.SMZ3.Tracking/Tracker.cs:892]: This one, too.

**Review by Vivelin — APPROVED** [review]

**Pull request merged**